src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs

```diff
@@ -0,0 +1,159 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace System
+{
+    public static partial class Convert
+    {
+        // Based on "Base64 encoding with SIMD instructions" article by Wojciech Muła
+        // http://0x80.pl/notesen/2016-01-12-sse-base64-encoding.html
+
+
+        // SSSE3-based implementation
+
+        private static unsafe void EncodeBase64Sse(byte* input, int inputLength, byte* output, int outputLength)
+        {
+            byte* outputCurrent = output;
+            int stride = 4 * 3;
+
+            Vector128<byte> shuffleMask = Vector128.Create((byte)1, 0, 2, 1, 4, 3, 5, 4, 7, 6, 8, 7, 10, 9, 11, 10);
+            Vector128<byte> shiftLut = Vector128.Create(18229723555195321415UL, 72503040736508UL).AsByte();
+            Vector128<byte> asciiToStringMaskLo = Vector128.Create(0, 0x80, 1, 0x80, 2, 0x80, 3, 0x80, 4, 0x80, 5, 0x80, 6, 0x80, 7, 0x80);
+            Vector128<byte> asciiToStringMaskHi = Vector128.Create(8, 0x80, 9, 0x80, 10, 0x80, 11, 0x80, 12, 0x80, 13, 0x80, 14, 0x80, 15, 0x80);
+
+            int i = 0;
+            for (; i < inputLength - stride - 1; i += stride)
+            {
+                Vector128<byte> inputVector = Sse2.LoadVector128(input + i);
+                inputVector = Ssse3.Shuffle(inputVector, shuffleMask);
+
+                Vector128<byte> t0 = Sse2.And(inputVector, Vector128.Create(0x0fc0fc00).AsByte());
+                Vector128<byte> t1 = Sse2.MultiplyHigh(t0.AsUInt16(), Vector128.Create(0x04000040).AsUInt16()).AsByte();
+                Vector128<byte> t2 = Sse2.And(inputVector, Vector128.Create(0x003f03f0).AsByte());
+                Vector128<byte> t3 = Sse2.MultiplyLow(t2.AsUInt16(), Vector128.Create(0x01000010).AsUInt16()).AsByte();
+                Vector128<byte> indices = Sse2.Or(t1, t3);
+
+                Vector128<byte> result = Sse2.SubtractSaturate(indices, Vector128.Create((byte)51));
+                Vector128<sbyte> compareResult = Sse2.CompareGreaterThan(Vector128.Create((sbyte)26), indices.AsSByte());
+                result = Sse2.Or(result, Sse2.And(compareResult.AsByte(), Vector128.Create((byte)13)));
+                result = Ssse3.Shuffle(shiftLut, result);
+                result = Sse2.Add(result, indices);
+
+                // save as two-bytes string
+                Sse2.Store(outputCurrent, Ssse3.Shuffle(result, asciiToStringMaskLo));
+                outputCurrent += Vector128<byte>.Count;
+                Sse2.Store(outputCurrent, Ssse3.Shuffle(result, asciiToStringMaskHi));
+                outputCurrent += Vector128<byte>.Count;
+            }
+
+            // Handle cases when inputLength is not a multiple of 24
+            if (i != inputLength)
+            {
+                ConvertToBase64Array((char*)outputCurrent, input, i, inputLength - i, false);
```

Comment:
Is this a case where, for the final iteration, you could do a back-track and mask the already processed elements?

Reply:
um not sure I understand what you mean - if i != inputLength it fully forwards data to `ConvertToBase64Array` to complete and does `return;` if `i == inputLength` but inputLength % 3 != 0 - we need to insert paddings - we can ask `ConvertToBase64Array` to do it (see AVX impl) or insert it here (for better perf?)

Reply:
This looks like its handling trailing elements...

In a number of vectorized algorithms, it is possible to handle the trailing elements still with vectorization. This is accomplished by backtracking so that the final read is exactly 16-bytes, and then masking out any elements that were already processed. For example, we do that with `Scale` (multiplying a matrix by a scalar value) here: https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.CpuMath/SseIntrinsics.cs#L473

I was wondering if this algorithm would allow the same, as it (the additional couple instructions for masking) is often cheaper than looping over the remaining elements in scalar.

Reply:
@tannergooding there can be a case when I have a final `Vector256<byte>` (32 bytes) and there is no room for it in the output string to do `Avx2.Store`. It can crash with `ExecutionEngineException`.
I tried to remove this trailing elements processing and do everything on vectors and it worked! but with a chance of the crash 🙂

Reply:
> and then masking out any elements that were already processed

Masking isn't needed, as the algorithm is idempotent, so you can process the already processed elements again (yielding the same encoded results).

But in my tests the backtracking introduced more overhead (especially to handle padding in a correct way, with setting a pad-vector and or-ing the result), so that a simple loop for the final elements resulted in better performance.